### ISSUE TITLE
Django 3.0 support

### ISSUE THREAD

**Issue by dubirajara** [radiac/django-tagulous#85]
Hi, 
six is not available in django.utils in Django 3 ,  all python 2 support is dropped.

thank you

**Comment by MrAlexWinkler** [comment]
I had the same issue and found an answer here
https://stackoverflow.com/questions/59193514/importerror-cannot-import-name-six-from-django-utils

Looks like "six" was depreciated from django.utils
https://docs.djangoproject.com/en/3.0/releases/3.0/#removed-private-python-2-compatibility-apis

This is a package that replaces it
https://pypi.org/project/six/

Wherever "six" was you can replace it with PY3
https://six.readthedocs.io/

Make sure to replace
from django.utils import six
with
from six import *

The reason all and not just PY3 is because python_2_unicode_compatible also needs to be imported.

**Comment by radiac** [comment]
Thanks for flagging this. Django 3 has been out for a couple of weeks now and it's definitely on my radar; I don't have much time to tackle it at the moment, but I'm hoping to look at it soon.

My thoughts in more detail:

The advice for package maintainers is to drop support for all versions of Django prior to 2.2 (see [3.0 release notes](https://docs.djangoproject.com/en/3.0/releases/3.0/#third-party-library-support-for-older-version-of-django)) - that sounds sensible and is really what I'd like to do as it would simplify maintenance, but I know of several projects on pre-2.2 which are waiting on some other tickets to be closed. Ideally I'd sort those and put out a final release for pre-2.2, then a new release dropping pre-2.2 and supporting 3.0.

I'm hoping I can clear some of my other work over the Christmas break and get to this early in the new year, but if it doesn't happen soon I'll look at putting out a temporary release which adds 3.0 support without losing pre-2.2.

(When I say pre-2.2 I really just mean 1.11 LTS, but even that is only supported til April, so a stop-gap which keeps pre-2.2 support til then doesn't seem terrible) 

I'll post on this issue whenever there's some news about a release which support 3.0.

**Comment by MrAlexWinkler** [comment]
I believe I'm having an issue with my solution above. Getting an error message with

`    elif isinstance(value, PY3.string_types):
AttributeError: 'bool' object has no attribute 'string_types'
`

will update if I find any solution been stuck on it for a bit.

**Comment by ensignavenger** [comment]
Hi!  Wondering what the status of this is?  I created my own local fork for the time being, but would rather use an official release.

PS, Thanks for creating such a fabulous project!

**Comment by jwsmithers** [comment]
Hi @ensignavenger , would you be willing to share your fork until the fix is official?
I'd like to try this project, but am limited to django3.

**Comment by ensignavenger** [comment]
@jwsmithers- apologies for not responding sooner- the project I am using this in is a side project, so I only get to work on it here a little and there a little.  I also vendored the project into my main project, so it isn't just in a repo I can make public... but it shouldn't be too hard for me to extract it back out and upload it to a public repo.  I will try to do that some time this week.  If you don't hear back from me before then, please feel free to ping me again :)  Like I said, it is  aside project that gets attention here and there as I have time.

**Comment by jwsmithers** [comment]
@ensignavenger , no worries, thanks for the response! 
If you have time for that it would be absolutely fantastic! 
I had a look through the fixes mentioned above but it seems to be a little more involved. Specifically with reference to the `python_2_unicode_compatible`, but maybe I'm just missing something.

**Comment by ensignavenger** [comment]
@jwsmithers I just removed all backwards compatibility, because I don't need it.  I am only using Python3.8 with Django 3.0, on this project.  Here is my a repo where you can get my code: https://github.com/ensignavenger/tagulous

I hope that helps.

**Comment by jwsmithers** [comment]
@ensignavenger , can't thank you enough 😄

**Comment by changeling** [comment]
@radiac Any word on 3.0 support?

**Comment by radiac** [comment]
This is still my main priority in my free time, but I've not had as much of that as I'd have liked this year! I put out v1 last week which resolves most serious outstanding issues for <=2.2, so I've started on the next version which will have full dj3 support - I'm stripping out pre dj2.2 code, then it should just be a case of a few tweaks to make sure the tests pass.

I can't give an exact estimate yet, but soon™ - the end is definitely in sight!

Thanks everyone for your understanding and support.

**Comment by ensignavenger** [comment]
Thank you @radiac for making such a fantastic tagging module and licensing it as Open Source!  Very glad to hear that a dj3 compatible version is coming!

**Comment by changeling** [comment]
Really looking forward to it! I was just checking in, as I'm starting a project and wrestling with "one of the other ones," so I thought I'd see if tagulous was ready. Next project, I'm looking forward to using it. Cheers!

**Comment by radiac** [comment]
I've pushed a branch which supports Django 2.2, 3.0 and 3.1 on Python 3.5+ (`feature/django3.1`). There are a lot of cleanup changes in here (stripped out py2, south (!), django 2.1 and earlier) - the tests all pass, but I want to run through a few more manual tests over the next week or so before I push it out to pypi.

If anyone would like to beta test it in the meantime, you should be able to install it with:

    pip install -e git+https://github.com/radiac/django-tagulous.git@feature/django3.1#egg=django-tagulous

**Comment by changeling** [comment]
@radiac Thanks for this! I'm testing it on a project I'd put taggit on, but currently implementing it in parallel because the migration path in the docs is failing with the 'tags_for' bug and the posted workaround on their site doesn't seem to work. That might be something to re-write in your copious free time. :D

**Comment by jedie** [comment]
Is there any release date?

And: Why is the django version not restricted in "setup.py" ?

**Comment by leonh** [comment]
Hi, I I'm interested in switching to django-tagulous in a new Django 3.1 project. Is Django version 3 support pretty much completed now?

**Comment by changeling** [comment]
@leonh I've been using the 3.1 branch in a project for a little while now, and haven't had any issues. Just for the record.

**Comment by Rohail1** [comment]
Hello, I have upgraded django to version 3.1.4 and now I wanted to install the upgraded django-tagulous but I am getting error when I run the following code. I have attached the image of the error I am getting

`https://github.com/radiac/django-tagulous.git@feature/django3.1#egg=django-tagulous`
![image](https://user-images.githubusercontent.com/5876740/101283411-f7e17900-37fb-11eb-9e3a-12b41c2f4b05.png)

**Comment by radiac** [comment]
Dj3.0 and 3.1 support have been released for a while, and I'm about to release support for dj3.2, so I'm going to close this. If anyone is still having problems please create a new issue.
